app/segments/drill/[id]/submission/input.js

```diff
@@ -136,7 +153,114 @@ async function uploadAttempt(
   }
 }
 
-//TODO: Create a function to check leaderboard and update if needed
+//A function to check leaderboard and update if needed
+function handleLeaderboardUpdate(
+  uploadData,
+  drillInfo,
+  currentLeaderboard,
+  teamId,
+  drillId,
+  userId,
+  queryClient,
+) {
+  const leaderboardData = currentLeaderboard;
+  const invalidateKeys = [
+    ["user"],
+    ["drillInfo"],
+    ["best_attempts", teamId, drillId], // keep teamId param as it is a string argument in useLeaderboard's query key
+    ["attempts", { drillId }], // stats pages
+    ["attempts", { userId }], // for profile index (list of drill types)
+  ];
+
+  //check if the user exists on the leaderboard
+  if (leaderboardData[uploadData.uid] == undefined) {
+    console.log("User not on leaderboard, uploading this attempt");
+
+    uploadNewLeaderboard(
+      drillInfo.mainOutputAttempt,
+      uploadData.uid,
+      uploadData.did,
+      uploadData[drillInfo.mainOutputAttempt],
+      uploadData.id,
+    );
+    // invalidate cache after successful upload
+    invalidateMultipleKeys(queryClient, invalidateKeys);
```

Comment:
This should be handled in the submit function or right after it (in the submit button maybe). This way you don't have to pass queryClient and teamId to this function

Reply:
Or if this is to be handled like completeAssigned, then there should be an await here, and this function should be async

Reply:
I moved the "invalidateMultipleKeys" hook call into the handleButtonClick function (submit function) as suggested, and removed the queryClient and teamId inputs I was passing into handleLeaderboardUpdate / uploadNewLeaderboard functions.

Reply:
There is a bit of a lag for the updated value to show up on the leaderboard (so maybe can explore async in future) but for now I think it works ok.